File: ios/FamilyApp/App/RootViewModel.swift
// Auth gate — the iOS twin of RootViewModel.kt. Emits Loading → SignedOut → SignedIn
// based on the stored app user id (public.users.id, NOT the auth UUID).
import Foundation
import Observation

enum AuthGate {
    case loading
    case signedOut
    case needsPermissions
    case signedIn
}

@Observable
@MainActor
final class RootViewModel {
    private let store = SessionStore.shared
    private let repo: FamilyRepositoryProtocol
    private var bootstrapped = false
    private var didSyncAfterSignIn = false

    init(repo: FamilyRepositoryProtocol = FamilyRepository.shared) {
        self.repo = repo
    }

    var gate: AuthGate {
        if !bootstrapped { return .loading }
        if store.currentUserId == nil { return .signedOut }
        if !store.permissionsRequested { return .needsPermissions }
        return .signedIn
    }

    /// Restores the Supabase auth session (supabase-swift persists it in the keychain),
    /// then opens the gate. The stored app user id is the source of truth — same as
    /// Android's DataStore flow.
    func bootstrap() async {
        guard !bootstrapped else { return }
        // Touch the session so token refresh happens before the first query.
        _ = try? await SupabaseClientProvider.client.auth.session
        bootstrapped = true
        if store.currentUserId != nil {
            await repo.touchLastActive()
        }
    }

    /// Once fully signed in, register this device's push token and mirror the local
    /// notification settings to the server (Android RootViewModel parity).
    func onSignedIn() async {
        guard !didSyncAfterSignIn else { return }
        didSyncAfterSignIn = true
        await repo.syncPushToken()
        await repo.syncNotificationPrefsToServer()
    }

    func completePermissionsOnboarding() {
        store.setPermissionsRequested()
    }
}

File: android/app/src/main/java/com/example/mainactivity/ui/navigation/RootViewModel.kt
package com.example.mainactivity.ui.navigation

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.example.mainactivity.data.FamilyRepository
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.flow.SharingStarted
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.combine
import kotlinx.coroutines.flow.stateIn
import kotlinx.coroutines.launch
import javax.inject.Inject

sealed interface AuthGate {
    data object Loading : AuthGate

    data object SignedOut : AuthGate

    data object NeedsPermissions : AuthGate

    data object SignedIn : AuthGate
}

@HiltViewModel
class RootViewModel
    @Inject
    constructor(
        internal val repo: FamilyRepository,
    ) : ViewModel() {
        val gate: StateFlow<AuthGate> =
            combine(
                repo.currentUserId,
                repo.permissionsRequested,
            ) { userId, permsDone ->
                when {
                    userId == null -> AuthGate.SignedOut
                    !permsDone -> AuthGate.NeedsPermissions
                    else -> AuthGate.SignedIn
                }
            }.stateIn(viewModelScope, SharingStarted.WhileSubscribed(5000), AuthGate.Loading)

        init {
            // Once fully signed in, register this device's FCM token and mirror the local
            // notification settings to the server so push delivery works and respects them.
            viewModelScope.launch {
                gate.collect { state ->
                    if (state is AuthGate.SignedIn) {
                        repo.syncPushToken()
                        repo.syncNotificationPrefsToServer()
                    }
                }
            }
        }

        fun completePermissionsOnboarding() =
            viewModelScope.launch {
                repo.setPermissionsRequested()
            }
    }
